### DISCUSSION TITLE
OxygenOS 12 Bluetooth Bug

### DISCUSSION THREAD

**Discussion by Miquest** [NightscoutFoundation/xDrip#2106] [Q&A]
Hello everyone,
I updated my OnePlus 8T to Android 12, but now the OB1 Collector goes always in Deep sleep and it has a long WakeUp time (5min).
3 Readings, and one missing. And then again...
It's confusing me...
Has anyone a solution?

Greetings,
Thorin

**Comment by Navid200** [comment]
If it is a G6, do all settings agree with the following?
https://navid200.github.io/xDrip/docs/G6-Recommended-Settings.html

**Reply by Miquest** [reply]
Yes, I read this Instructions too...
But it's not working.
It worked very well on OxygenOS 11.
I think, Android is blocking Xdrip, or is shutting down the App in the Background.
I need an Option to force the Background work!

The longest Wakeup Time is 10min...

**Comment by Navid200** [comment]
@jamorham

**Reply by Miquest** [reply]
@jamorham 

Maybe you can add an option to force Restarting of the OB1 Collector 30 Seconds before the Next Reading...

If I restart the Collector Manually, it worked

**Comment by Miquest** [comment]
Any other Idea to fix that?

**Reply by Navid200** [reply]
What do you see if you go to 
Android Settings &#8722;> Apps & notification &#8722;> See all *** apps 

Then, from the 3-dot menu, tap on "Show system apps".

Scroll through and tap on "Bluetooth".
Now, tap on Advanced.

What do you see?

**Comment by Miquest** [comment]
![Screenshot_2022-05-12-14-53-37-00_0ba066473b79d6e213a1f6f52505e2ee](https://user-images.githubusercontent.com/91971937/168079212-90ce3a22-180a-47e6-b473-714015975c4b.jpg)
![Screenshot_2022-05-12-14-52-59-36_fc704e6b13c4fb26bf5e411f75da84f2](https://user-images.githubusercontent.com/91971937/168079214-449ba164-1329-47f8-a040-b13de8ded0af.jpg)

**Reply by Miquest** [reply]
I can See only this options.
There is no option to turn off the Battery Optimization like in Android 11.
Maybe it's a bug in OxygenOS.

**Reply by Navid200** [reply]
Is that under battery?

**Reply by Miquest** [reply]
Yes, the first Screenshot.
The Second is under App Management

**Reply by Miquest** [reply]
Settings<Apps<App Management

In the Battery Optimization Settings I can only see the Consumption of the Bluetooth Function

**Reply by Navid200** [reply]
On some phones, instead of battery optimization, you get the option of selecting unrestricted.
That is specially the case for Pixel.

**Reply by Miquest** [reply]
I activated this option.
I allowed activity in Background and automated Start in the background.

In the Error log is a Noise Warning:
Startup to low

Maybe that's helpful?

**Comment by Miquest** [comment]
I have the Nightly xDrip built from yesterday installed, because I hoped that there is a optimization of my error in it.

**Comment by Navid200** [comment]
There are many who have the intermittent connectivity problem.
We already have multiple open issues.  
These are just two examples.
https://github.com/NightscoutFoundation/xDrip/issues/1664
https://github.com/NightscoutFoundation/xDrip/issues/1789

We have so many closed issues.  Someone can permanently delete them.  But, that takes time.
We still need to be able to search through closed issues.  Having too many of them makes our job harder.

That's why we have asked everyone to use discussions instead.  If an item comes up that is persistent and we cannot address it in discussions and needs to be on the 4record as an issue, we then ask the user to open an issue.

This subject does not fit that category because we already are aware of it.

Every device has different ways of stopping processes that run in the background.  xDrip needs to run in the background.
For every phone, one needs to find the way to disable all means in play that could interrupt xDrip.

I'm not saying your problem is not real.  I'm not saying it is not annoying.  But, opening another issue is not going to help.  And closing more and more issues that you open does not permanently delete the,

I respectfully ask you to stop opening issues.

**Reply by Miquest** [reply]
I am so sorry, i don't want to bother you.
I wanted to edit one reply I wrote, but I opened another issue.
I know, I shouldn't open other issues, that was not my purpose

**Reply by Miquest** [reply]
I read the first post now, and I added the OB1 Debug tag for the log.
Thank you for the great support and the quick answers.
I'll wait for an Update that is functioning :)

**Comment by Miquest** [comment]
@jamorham, 
Maybe you could add a function, named "Change WakeUp Time" or something like that, which can change the WakeUp Time of the OB1 Collector?
I think, that could be helpful because I have a long wakeUp Time, around 2 Minutes.
So I could change the Time of WakeUp before next reading


![Screenshot_2022-05-13-20-57-33-66_40deb401b9ffe8e1df2f1cc5ba480b12](https://user-images.githubusercontent.com/91971937/168363200-0736d787-af01-4702-a7e9-925db34afcaf.jpg)

**Comment by Navid200** [comment]
When you see the highlighted in red note on the system status page "slowest wakeup: ", that's an effect, not a cause.
If xDrip is not allowed to operate freely in the background, there is not much we can do by changing xDrip itself.  If Android doesn't let xDrip wake up, xDrip cannot wake itself up.  xDrip is an app.  Android is the operating system.
It's like pulling up on your own shoelaces hoping to lift yourself up from the ground.

**Reply by Miquest** [reply]
I know what you mean, but you could possibly counteract the delayed WakeUp by setting a WakeUp Time earlier

**Reply by Navid200** [reply]
I see what you mean now.  Thanks for explaining.
Sorry.  Jamorham is the guy to respond to that.

**Reply by Miquest** [reply]
No problem.
I think, Jamorham has also other Things to do...

Thank you so much for your quick answers, I never had a Online Discussion like that :)

**Reply by Navid200** [reply]
I will add that after Android workarounds was removed out of necessity, we now have other concerns.  For example, xDrip cannot scan for too long.  Or else, it could cause further issues with connectivity.
That's why Avoid scanning was added.  We now ask everyone to enable both minimize scanning and avoid scanning for Android 10 and above.

Waking up the scanner too early may contradict those.
But, again, these are all things I have learned listening to jamorham.  The real detail is actually too complicated for me.

**Reply by Miquest** [reply]
I activated both options.
Android Workarounds are also enabled.
I have no experience in Java, only in Python and JavaScript...
I can only read the code and make suggestions...
But thank you for adding this.

**Reply by Miquest** [reply]
Yes, if xDrip is scanning too long, that could be problematic, with other Bluetooth devices I think...

**Reply by Navid200** [reply]
How/where did you enable Android Workarounds?  Please explain what you mean.  I am not aware of such an option.
What version of xDrip are you using?

**Reply by Miquest** [reply]
I'm using the newest build of xDrip.

Enabled in the Debug Settings are:
Constant Scanning,
UI-Thread
Full Authentication
OB1 can initiate a bond
![Screenshot_2022-05-13-22-09-42-88_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/168382330-7d352129-378a-49fd-95d1-830a90380548.jpg)
![Screenshot_2022-05-13-22-09-28-91_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/168382333-8503b23a-9187-42f4-bda2-4f84102a4d67.jpg)
![Screenshot_2022-05-13-22-09-15-93_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/168382335-f81b3d33-67b1-41b3-a94c-239fff099518.jpg)

**Reply by Navid200** [reply]
I see no Android workarounds.

Anyway, this is where I ask everyone to start from:  https://navid200.github.io/xDrip/docs/G6-Recommended-Settings.html

Your settings are different.  For example, i don't know why you would have "Scan for G5 constantly set" if that's what you mean by constant scanning.

I would be interested to know how everything goes if you use those settings in my link.

**Reply by Miquest** [reply]
I changed my Settings like in the link, but there are still missing readings.
I turned off the Battery Optimization for Xdrip, the Location permission is granted every time, and Bluetooth Battery Optimization doesn't exist...

**Reply by Miquest** [reply]
I activated "Scan for G5 constantly", because in my experience, it's working better.

**Reply by Navid200** [reply]
But, that's contradictory with minimize scanning.
With all settings as shown, would you please show the logs for when a disconnect occurs?

**Comment by Miquest** [comment]
My log with the recommended Settings

![Screenshot_2022-05-14-22-01-55-01_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/168446654-5a59494d-bdd6-4fca-bcb9-5f958991d91d.jpg)

**Reply by Navid200** [reply]
@jamorham Would you please have a look at this?
@thorinEichenschild007 Please enable detail logs as explained here:  https://github.com/NightscoutFoundation/xDrip/wiki/Enabling-Debugging-for-OB1-collector

**Reply by Miquest** [reply]
I enabled the OB1 Debugging before I sent the Picture.

**Comment by Miquest** [comment]
Maybe that's helpful?

![Screenshot_2022-05-16-05-49-44-93_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/168516731-65d296df-4750-47d9-b01e-203f96b1075c.jpg)

**Comment by Miquest** [comment]
Update:
The Nightly Build from the 19th May is able to read every second reading...
That's okay, but not perfect...

**Reply by Navid200** [reply]
That's odd.  There is no change I can see that would impact your transmitter in that release.
Are you sure the May 19 and May 16 releases behave differently with respect to this problem?

**Reply by Miquest** [reply]
No, unfortunately they don't... But I install each new build hoping it solves my problem

**Reply by Miquest** [reply]
![Screenshot_2022-05-21-15-39-14-60_2dd2c1bcb8f659cd696bcba2072f751b](https://user-images.githubusercontent.com/91971937/169654254-d1cc3ca3-fc0f-4580-adb8-7bd8bca4fd19.jpg)

The Device is not able to connect to the G6 Transmitter

**Comment by Navid200** [comment]
Have you seen this?
https://github.com/NightscoutFoundation/xDrip/issues/1664

**Reply by Miquest** [reply]
Yes, I read this post...
And I switched Back to your recommended Settings.
Now, it's working.

**Comment by m-l-s0** [comment]
Have you found a solution yet, I updated to Oxygen OS 12.1 (which is supposed to be stable :/) two days ago and have been having problems with Xdrip not connecting to the G6 when the screen is off ever since.
I suspect the System is killing the bluetooth service in the background.
The xDrip logs say `Agressively restarting collector service due to lack of reception:backoff: 480 ` where 480 is increasing by 60 every message.

**Reply by Miquest** [reply]
My solution is the Dexcom BYODA. 
I installed it and it's working as collector now, because I don't like the Dexcom App.
So, I use xDrip as main application and the BYODA only as collector.
See the Android APS Documentation for instructions!

**Reply by Miquest** [reply]
It's only a workaround, but on my new Find X5 Pro (ColorOS 13, 12 before) it's working without any issues!

Edit, Link to Documentation:
https://androidaps.readthedocs.io/de/latest/Hardware/DexcomG6.html

**Comment by andyrozman** [comment]
I used the settings as proposed and results are even worse than before... I am on OnePlus 8 with Android 13 (Oxygen 13). I use the lates nightly build.

**Reply by Miquest** [reply]
I'm using the BYODA Version of the G6 Application with Broadcast to xDrip Plus to avoid this problem. It's working perfectly. 
My device is the Oppo Find X5 Pro with ColorOS 13.

**Reply by andyrozman** [reply]
Was trying to do the same, but BYODA can't find my transmitter... I will retry after I start new session and new transmitter (in 6 days)... But this is more of workaround than real solution.

**Reply by andyrozman** [reply]
@Navid200 
With proposed changes I was getting "133" errors (133 GATT ERROR bug). So I am trying to revert changes one by one...

**Reply by Navid200** [reply]
This thread has become too big.  There are so many posts in it and some people like to attach very large images.  All of that makes it hard for anyone interested to find out what the problem really is and the title starts with OxygenOS.  I don't even know what that means.

Can you summarize in a new thread?

**Reply by andyrozman** [reply]
Should we create discussion instead, because I think any of people using OnePlus with one of new Android releases (12 or higher), will be affected?
OxygenOS is packaging of their own applications and everything vendor specific based on Android version, so OxygenOS 12 is A12 version and 13 is A13 version...

**Reply by Navid200** [reply]
> Should we create discussion instead, 

I don't understand what you are asking.
I have asked you to if you don't mind please start a new discussions including in it only the minimum so that JamOrHam for example can have a look and find out what it is all about.

I haven't asked you to open an issue.  Please don't do that yet.

**Reply by m-l-s0** [reply]
is Color OS based on Oxygen OS? Bacause since thorin has the same problem on a different ROM it might be an issue with the OnePlus-devices/drivers ans Android 12+ ?

I'm gonna see if I can get it to run on my OnePlus One runing Lineage OS tonight.

**Reply by Miquest** [reply]
OxygenOS 12/13 and ColorOS 12/13 are based on one and the same software kit from Oppo, so the architecture is almost identical.

**Reply by Miquest** [reply]
@andyrozman,
Did you completely decouple the transmitter and then install the BYODA? Everything went well for me. Sure, it's a workaround, but a well-functioning one that doesn't have any negative consequences for me.
Battery drain doesn't exist on my phone, xDrip and Dexcom are consuming 3% of my battery in 24 hours...

**Comment by Navid200** [comment]
@andyrozman Do I understand correctly?

There are phones that can use xDrip fine.  But, users flash the ROM and xDrip has problems and we are being asked to change xDrip to address that?!

**Reply by andyrozman** [reply]
No. When you buy OnePlus you get it installed with OxygenOS which is made by OnePlus and is based on certain Android image... So Oxygen 12 contains Android 12 and so on... 

So this is original software for this phones... Problem is that with default (proposed) settings Xdrip doesn't work (good)... I have disabled not only optimization for Xdrip, I i disabled power optimization for whole phone...  So with certain settings Xdrip was working, but every 2nd or 3rd reading would be missing, so alltogether realtime capture at 60-65%.

**Reply by andyrozman** [reply]
My old phone was OnePlus 5T and I used different version of OxygenOS with it (stated with 8.1 and then continued with 9 and last update was 10) and Xdrip worked good on thoose... But it seems that with 12 something changed and now it doesn't work so good anymore... (I noticed that only when I switched phones, because my old phone's highest upgrade version was 10).

**Comment by jamorham** [comment]
If you guys are getting wakeup jitter > 10 seconds that that will be what stops it working. This happens if the phone android framework does not meet the android specifications due to aggressive non-standard power saving.

I just added a fix today for Redmi Android 12+ phones which have similar problems that we had seen only on samsung previously. https://github.com/NightscoutFoundation/xDrip/commit/b3ba148725f2b3806f7ac79dd90a2ef7b641aed0 

If you are self compiling xDrip then you could make a similar adjustment for your phone model/manufacturer or please feel free to send me those details (submitted logs include this for me) and I can produce a custom test version for you. But this is only if you regularly see wakeup jitter > 10 seconds recorded in the logs or on the status page.

**Reply by Navid200** [reply]
Dumb question:

What would be the disadvantage of this?
In other words, why not make this procedure default for all manufacturers?

**Reply by jamorham** [reply]
On some devices it will display icons showing that an alarm clock is set and on others it will report things like, "next alarm clock in 5 minutes" which can interfere with the user's normal use of the android alarm clock features (but only cosmetically)

**Reply by jamorham** [reply]
But sorry yes, we could make it automatically try this approach based on heuristics if there is significant wake up jitter but as this had previously only been a problem on samsung I restricted the workaround to those devices.

**Reply by andyrozman** [reply]
Yes. I am getting jitter from 13 seconds up to 4 minutes in some cases... But some of thoose might be because of "proposed settings" I applied in the morning (when I saw this bug).

I disabled all power savings for whole phone... But there still might be some in background... I also set "Allow background activity to true".

**Reply by Miquest** [reply]
I'm using the Macrodroid App, which is also using the system clock (alert app) for some features.
I disabled the icon with the clock in the status bar and that's it... 
It's working good!

**Reply by Miquest** [reply]
Could you add it in the Debug Options for G5/G6?
If people need that feature if they have no Samsung or Xiaomi phone, they can enable it.

**Reply by andyrozman** [reply]
@jamorham Would jitter mean that somehow "scheduler", missed the reading window. I don't know how exactly arhitecture works for G6 collections, but would moving scheduler into UI thread help with that, or into foreground service? Then time/windows wouldn't be missed so easily? 

I have both "Allow foreground activity" and "Allow background activity" checked and it still doesn't help much...

**Reply by jamorham** [reply]
@andyrozman those are good suggestions but the scheduler shouldn't miss its window because its part of the framework and has a contract for when it should wake up. Its due to manufacturers breaking this contract with power saving. I have added oneplus to today's nightly for the workaround so please let me know if that helps with the situation.

**Reply by Miquest** [reply]
@jamorham , great!
I hope it works!
Could you also add Oppo to this workaround in the next nightly release? I think they are also affected of the problem, because on my Oppo phone there are the same problems as in my Oneplus phone, a high jitter. ColorOS has almost the same architecture as OxygenOS, so think we can avoid the problem in the same way.
I would test it on my Oppo Phone as soon as I have replaced the sensor (in 5 days) to make sure that I can also use the BYODA if doesn't work as I thought.

**Reply by andyrozman** [reply]
@thorinEichenschild007 @jamorham 
So tested changes and it works great... I have 97% realtime capture rate... 

I still have all powersaving features disabled and I use some of my own configuration... I will enable powersaving (of phone) after midnight, and will keep it running that way, and if everything works ok, I might see if I can change settings towards recommended ones. I am not sure if all of them will work, we might need special page with recommended settings for this phone, but I have few more days of testing before me... 

But so far results are very good.

**Reply by andyrozman** [reply]
So results from yesterdays test... I have enabled phones power saving and capture rate is still good (100% actually, my first test probably little skewed, because I started at 1 AM not at 0).

Today I will try to see if I can test different settings, that I use and are not recomended, to see if any of them is not working for us.

**Reply by VoronMods** [reply]
I also have OnePlus 9 Pro with OxygenOS 13 (11->13, skipped 12) installed two days ago, first saw the "Samsung Workaround" enabled on latest nightly and went back to old xDrip+ version, after reinstalling latest (before I saw this post here) and finding all the power saving stuff, finely got 98% capture rate and no connection issues, before maybe every 10 min got BG data with Dexcom G6
The "Samsung Workaround" got me confused as I tough "this is not a Samsung here what I got" and after I could not disable it, went for few hours back to old xDrip+ version.

**Reply by VoronMods** [reply]
> So results from yesterdays test... I have enabled phones power saving and capture rate is still good (100% actually, my first test probably little skewed, because I started at 1 AM not at 0).
> 
> Today I will try to see if I can test different settings, that I use and are not recomended, to see if any of them is not working for us.

Which setting do you have enabled/disabled? Can you post screenshots of your xDrip+ and OxygenOS 13 settings? (here or on Discord)

**Reply by Miquest** [reply]
@jamorham,
did you add also Oppo to the Workaround feature?
Seems as if it's working well, so I would be happy to be able to use it also on my Find X5 Pro!
I can give feedback for that on Tuesday when I replace my G6 Sensor

**Comment by andyrozman** [comment]
@VoronMods 
So on OxyGen 13 I have enabled all power saving, but for Xdrip is on "Not optimize". And on power consumption controls I have "Allow background activity" enabled.

For Xdrip you can set all recommended settings. And so far only difference I have is following options (and I am still testing if this is needed). So I have:
-  "Force G5 to UI Thread" enabled
- "Dex BT Watchdog" disabled
- "Trust auto connect" disabled

So I am still testing this last 3 settings... Just to see what is needed...

**Reply by VoronMods** [reply]
> @VoronMods So on OxyGen 13 I have enabled all power saving, but for Xdrip is on "Not optimize". And on power consumption controls I have "Allow background activity" enabled.
> 
> For Xdrip you can set all recommended settings. And so far only difference I have is following options (and I am still testing if this is needed). So I have:
> 
>     * "Force G5 to UI Thread" enabled
> 
>     * "Dex BT Watchdog" disabled
> 
>     * "Trust auto connect" disabled
> 
> 
> So I am still testing this last 3 settings... Just to see what is needed...

on my, I have:
"Force G5 to UI Thread"  disabled
"Dex BT Watchdog" disabled
"Trust auto connect" enabled

I'm not aware of the default settings, or what they are, or if there is a checkbox to apply them.

**Reply by andyrozman** [reply]
There are not default settings, but there are recommended settings, which were posted upper in this discussion. 

And so far this 3 settings are the only difference and I am still testing if they are needed (plus using Less common settings -> Other misc -> Samunsg Workarround AND **using newest nightly version** )

**Reply by Navid200** [reply]
> There are not default settings, 

Yes, there are.  
The default settings are the settings that source wizard sets when you install xDrip for the very first time and use the source wizard.  The lead developer has chosen those.
The settings shown in my documentation, as the recommended settings, are those exact settings.

If you change your settings, as many users do before they post and ask for help, running source wizard after  will not undo your changes.

I would like to add a feature to xDrip to allow the user to return xDrip to default settings.   But, that can be tricky.  So, we have not done it yet.

**Reply by VoronMods** [reply]
> > There are not default settings,
> 
> Yes, there are. The default settings are the settings that source wizard sets when you install xDrip for the very first time and use the source wizard. The lead developer has chosen those. The settings shown in my documentation, as the recommended settings, are those exact settings.
> 
> If you change your settings, as many users do before they post and ask for help, running source wizard after will not undo your changes.
> 
> I would like to add a feature to xDrip to allow the user to return xDrip to default settings. But, that can be tricky. So, we have not done it yet.

haaaa, I installed my first xDrip+ maybe 6-7(?) years ago and dragging the settings since then :D

**Reply by Miquest** [reply]
@jamorham?
Did you add Oppo to the Workaround feature?
I'll replace my G6 Sensor now and could test the Workaround also for Oppo phones in the next days!

**Reply by andyrozman** [reply]
@jamorham 
So I was running Xdrip now with all recommended/default settings with exception of Samsung workarround option - which needs to be enabled (phone is now also in Power saving mode) for last week and everything seems to be working great. So this seems to be correct solution for Oxygen 12 and higher based systems (One Pluses with A12 and higher).

**Comment by VoronMods** [comment]
@andyrozman and all
I did some adjustment to xDrip+ settings in last few days as I'm still getting to know OxygenOS 13 on my OP9 Pro, and got to <1% xDrip battery usage (about 18h) and 98% capture rate, also running it in "Power saving mode".
In addition, I have AAPS/Ruffy running w/o connection issues.

For now I'm "happy" with this setting.
I'm on F.74 FW version

**Reply by Navid200** [reply]
Do you never turn the screen on on your phone?

**Reply by Miquest** [reply]
I'm not on the Workaround on my Oppo Find X5 Pro...
Did you add it to the Samsung Workaround?

**Reply by VoronMods** [reply]
> Do you never turn the screen on on your phone?

you mean me? Sure I do:
![2023-02-17 15 29 03](https://user-images.githubusercontent.com/18570447/219682654-000301ca-4f27-44d7-b8ef-1c1a561200cc.jpg)

**Reply by Navid200** [reply]
Every single time so far I have seen a reference to the xDrip high power consumption claim, it has been a misunderstanding.

Look at the following two screenshots from the same phone at the same time.
![Screenshot_20230217-092949](https://user-images.githubusercontent.com/51497406/219683233-fa639b82-1445-4753-a9e1-c15af3c2df5a.png)
![Screenshot_20230217-093026](https://user-images.githubusercontent.com/51497406/219683235-b5d38bd6-41d5-4b89-9b76-b699b1d9acba.png)

According to the first, xDrip uses 18%.  According to the other, only 1%.  The reason for the first one being wrong is that it excludes the mobile network, Bluetooth and screen.  

Under normal circumstances, there is no way xDrip would consume a considerable amount of power.  Of course if you have an unusual case where you never use the phone as a phone, you can end up in a situation where xDrip would be the main power hug.  But, xDrip is meant to be used on a regular phone that is also used for everything else we do.

**Reply by VoronMods** [reply]
> Every single time so far I have seen a reference to the xDrip high power consumption claim, it has been a misunderstanding.
> 
> Look at the following two screenshots from the same phone at the same time. ![Screenshot_20230217-092949](https://user-images.githubusercontent.com/51497406/219683233-fa639b82-1445-4753-a9e1-c15af3c2df5a.png) ![Screenshot_20230217-093026](https://user-images.githubusercontent.com/51497406/219683235-b5d38bd6-41d5-4b89-9b76-b699b1d9acba.png)
> 
> According to the first, xDrip uses 18%. According to the other, only 1%. The reason for the first one being wrong is that it excludes the mobile network, Bluetooth and screen.
> 
> Under normal circumstances, there is no way xDrip would consume a considerable amount of power. Of course if you have an unusual case where you never use the phone as a phone, you can end up in a situation where xDrip would be the main power hug. But, xDrip is meant to be used on a regular phone that is also used for everything else we do.

I use my OP 9 Pro as daily driver for everything.
I guess Motorola uses Battery usage different as OP does.
![Screen Shot 2023-02-17 at 16 07 28 PM](https://user-images.githubusercontent.com/18570447/219691349-d4ae9756-6251-4a05-acc3-ce99001abdb0.png)

**Reply by Navid200** [reply]
If you uninstall xDrip from your phone and you have to charge your phone only once every 3 days.  Then, you install xDrip and as a result you end up having to charge it every other day (once every 2 days instead of 3), we have a problem with xDrip, and we need to do something about it.  This is an example of xDrip power consumption making an impact on your life.

On the other hand, if you charge your phone every other day regardless of having xDrip on it or not, everything else I am afraid is just OCD.

In other words, does it really make a practical difference?  If not, forget it.  Life is too short.
